File: weird_string_case.py
def to_weird_case(string: str) -> str:
    cap = True
    word = ""
    
    for letter in string:

        print(letter, cap, word)
        if letter == " ":
            word += letter
            cap = True
        elif cap:
            word += letter.upper()
            cap = False
        else:
            word += letter.lower()
            cap = True

    return word


def main():
    string = "This is a test"
    expected = "ThIs Is A TeSt"
    got = to_weird_case(string)
    print(got == expected)
    print(got, expected)

if __name__ == "__main__":
    main()

File: weird_string_case.go
package main

import (
	"fmt"
	"strings"
	"unicode"
)

// my solution
func toWeirdCase(str string) string {

	var myStrArray []string

	for _, v := range strings.Split(str, " ") {
		myStr := ""
		for i, _ := range v {
			if i%2 == 0 {
				myStr += strings.ToUpper(string(v[i]))
			} else {
				myStr += strings.ToLower(string(v[i]))
			}
		}
		myStrArray = append(myStrArray, myStr)
	}
	return strings.Join(myStrArray, " ")
}

// Best Practice
func BestPractice(str string) string {
	cap := true
	var result []rune

	for _, c := range str {
		if c == 32 {
			result = append(result, c)
			cap = true
		} else if cap {
			result = append(result, unicode.ToUpper(c))
			cap = false
		} else {
			result = append(result, unicode.ToLower(c))
			cap = true
		}
	}
	return string(result)
}

func main() {
	got := toWeirdCase("Ab cD")
	expected := "Ab Cd"
	fmt.Println(got == expected)
	fmt.Println(got, expected)

	got = BestPractice("Ab cD")
	fmt.Println(got == expected)
	fmt.Println(got, expected)
}
